### ISSUE TITLE
GR32_LowLevel.Mirror PUREPASCAL does not produce expected values

### ISSUE THREAD

**Issue by andersmelander** [graphics32/graphics32#291]
The `GR32_LowLevel.Mirror*` functions are used in various places to clamp values to a given range. Notably, when sampling a bitmap, they are used to avoid out of bounds pixel coordinates.

For example, `Mirror(n, 5)` where n=[0..15] should return the values `012345432101234`.
In other words: `Mirror(n, Max)` should be a _triangle function_ with a wavelength of `Max`.

The current implementations has multiple issues:

- The algorithm of the `Mirror(Value, Max)` PUREPASCAL implementation does not match the algorithm of the assembler implementation. The PUREPASCAL implementation has incorrect handling of the edge conditions, so in the example above, instead of returning \
  `012345432101234` \
  it returns \
  `012345543210012`
- The `MirrorPow2` implementation does not special case negative values and appears to suffer from the same problem as the Mirror implementation.

**Comment by CuriousKit** [comment]
How goes the progress in fixing this?

**Comment by andersmelander** [comment]
I'm busy with other stuff so I'm not working on it (which is why I haven't assigned it to myself)

**Comment by CuriousKit** [comment]
That's fair.  I saw that you made a commit related to the bug and thought initially that the bug had been fixed, but a closer look at the commit messages indicates that it's not all fixed yet.  Still, these messages will indicate that the issue is still open.

**Comment by andersmelander** [comment]
For reference, what we need is a replacement for the current (broken) algorithm:

```delphi
function MirrorPow2(Value, Max: Integer): Integer; overload;
begin
  if Value and (Max + 1) = 0 then
    Result := Value and Max
  else
    Result := Max - Value and Max;
end;
```

**MirrorPow([0..13], 3)**
Expected: 0123210123210
Actual: 0123210**301232**

**Comment by CuriousKit** [comment]
I'll take a study of the file myself at some point.  I'm curious to know what the function is supposed to do exactly.

**Comment by andersmelander** [comment]
The `Clamp`, `Wrap`, and `Mirror` functions are primarily used when sampling bitmaps (e.g. resizing, rotating, transforming, etc). They specify how out of bounds pixel coordinates should be handled.
https://graphics32.github.io/Docs/Units/GR32/Types/TWrapMode.htm

**Comment by CuriousKit** [comment]
Ah I see, thank you.  Having played with textures in OpenGL, I can definitely see the importance of getting this function correct then!

**Comment by andersmelander** [comment]
...and also used by some gradient functions: https://graphics32.github.io/Docs/Units/GR32_ColorGradients/Classes/TCustomGradientPolygonFiller/Properties/WrapMode.htm

**Comment by andersmelander** [comment]
> I can definitely see the importance of getting this function correct then!

I would think so but the bug has been there since the functions introduction in 2007...

**Comment by lamdalili** [comment]
This  code gives the correct values :

values for range  0..15 
012345 432101 2345

values for range  -15..15
543 210123 454321 012345 432101 2345

```
function Mirror(Value, Max: Integer): Integer;
var
  DivResult: Integer;
begin
  Value := Abs(Value);
  DivResult := DivMod(Value, Max , Result);

  if Odd(DivResult) then
    Result := Max - Result;
end;
```

 Test code 
```
var
 i,n:integer;
 TxtOut:string;
begin
 n := 5;
 TxtOut := '';
 for I := 0 to 15 do
 begin
    TxtOut := TxtOut + Mirror(I, n).ToString;
    if((I + 1) mod (n + 1)) = 0 then // padding
      TxtOut := TxtOut +' ';
 end;
 Edit1.Text :=  TxtOut;
end;
```

**Comment by andersmelander** [comment]
@lamdalili Thanks but the Mirror function has already been fixed in the [LowLevel.Mirror](https://github.com/graphics32/graphics32/tree/LowLevel.Mirror) branch:

https://github.com/graphics32/graphics32/blob/a322add8db44d44d8fc445455cf05aefcee6d2b1/Source/GR32_LowLevel.pas#L1255-L1265

What remains to be fixed is the MirrorPow2 function:

https://github.com/graphics32/graphics32/blob/a322add8db44d44d8fc445455cf05aefcee6d2b1/Source/GR32_LowLevel.pas#L1321-L1330

See: https://github.com/graphics32/graphics32/issues/291#issuecomment-2016990612

**Comment by andersmelander** [comment]
I've had one more go at this problem and I have now come to the conclusion that a version of `Mirror(Value, Max)`, which takes advantage of Max being 2<sup>n</sup>-1, isn't possible. Therefore `MirrorPow2` will be deprecated.

Furthermore, after benchmarking the current implementation of `Mirror`, which uses `DivMod`:
```pascal
function Mirror(Value, Max: Integer): Integer;
var
  Quotient: Integer;
begin
  if Value < 0 then
    Value := -Value;
  Quotient := DivMod(Value, Max, Result);

  if ((Quotient and 1) = 1) then
    Result := Max - Result;
end;
```
where `DivMod` is defined as:
```pascal
function DivMod(Dividend, Divisor: Integer; out Remainder: Integer): Integer;
begin
  Remainder := Dividend mod Divisor;
  Result := Dividend div Divisor;
end;
```

I will be replacing that implementation with an iterative algorithm instead. Like this:
```pascal
function Mirror(Value, Max: Integer): Integer;
begin
  if Value >= 0 then
    Result := Value
  else
    Result := -Value;

  while (Result > Max) do
    Result := Abs(Max + Max - Result);
end;
```

Although I would normally reject an iterative algorithm like this, the `DivMod` is so expensive that the iterative solution outperforms it up to about Value > Max*30.

For the typical use case, where Mirror is used in resampling, we will never have more than one whole cycle so Value will never exceed 2*Max and here the iterative implementation is over twice as fast.
When used with gradients it can be more common to have more cycles but still far less than 30 cycles.

**Comment by CuriousKit** [comment]
That seems fair.  Unfortunately, Free Pascal's compiler is not yet clever enough to optimise ```DivMod``` - it's something I'm currently looking at.  Essentially it performs the integer division twice: once for the quotient and once for the remainder, on top of the procedure call and the like (although if ```DivMod``` has an assembly language implementation, this helps somewhat).

If ```Max``` is a known constant, you can make some compile-time optimisations and convert the divisions into a multiplication (something the compiler does automatically), but as you said, if you have some assumptions about the values of ```Value```, you can make some tailor-made optimisations.

**Comment by andersmelander** [comment]
>  if DivMod has an assembly language implementation, this helps somewhat

Yes, but `Mirror` using the assembler version of `DivMod` in `GR32_LowLevel`...
```asm
        PUSH      EBX
        MOV       EBX,EDX
        CDQ
        IDIV      EBX
        MOV       [ECX],EDX
        POP       EBX
```
...is still slower than the iterative solution.

I'm currently testing and benchmarking the new implementations of `Mirror` and expect to commit shortly. Annoyingly it turns out that my handwritten assembler versions isn't really any faster than what Delphi produces. Annoying because that means that it now is as fast as it gets (unless, of course, I've goofed somewhere).
```pascal
function Mirror(Value, Max: Integer): Integer; {$IFDEF FPC} assembler; nostackframe; {$ENDIF}
asm
{$IFDEF TARGET_x64}
        MOV       EAX, ECX      // Value
{$ENDIF}
        // EAX: Value
        // EDX: Max

        // Max2 := 2*Max
        LEA       ECX, [EDX+EDX]

        // Result := Value

@Loop:
        // Result := Abs(Result)
        TEST      EAX, EAX
        JNL       @Positive
        NEG       EAX


@Positive:
        // while (Result > Max) do
        CMP       EAX, EDX
        JLE       @Exit

        // Result := 2*Max - Result
        NEG       EAX
        ADD       EAX, ECX

        JMP       @Loop
@Exit:
end;
{$ENDIF}
```

Delphi's:
```
Benchmark.Mirror.dpr.11: begin
00B028B0 53               push ebx
00B028B1 8BDA             mov ebx,edx
Benchmark.Mirror.dpr.12: if Value >= 0 then
00B028B3 85C0             test eax,eax
00B028B5 7C04             jl $00b028bb
Benchmark.Mirror.dpr.13: Result := Value
00B028B7 8BC8             mov ecx,eax
00B028B9 EB04             jmp $00b028bf
Benchmark.Mirror.dpr.15: Result := -Value;
00B028BB 8BC8             mov ecx,eax
00B028BD F7D9             neg ecx
Benchmark.Mirror.dpr.17: while (Result > Max) do
00B028BF 3BD9             cmp ebx,ecx
00B028C1 7D10             jnl $00b028d3
Benchmark.Mirror.dpr.18: Result := Abs(Max + Max - Result);
00B028C3 8D041B           lea eax,[ebx+ebx]
00B028C6 2BC1             sub eax,ecx
00B028C8 99               cdq 
00B028C9 33C2             xor eax,edx
00B028CB 2BC2             sub eax,edx
00B028CD 8BC8             mov ecx,eax
Benchmark.Mirror.dpr.17: while (Result > Max) do
00B028CF 3BD9             cmp ebx,ecx
00B028D1 7CF0             jl $00b028c3
Benchmark.Mirror.dpr.19: end;
00B028D3 8BC1             mov eax,ecx
00B028D5 5B               pop ebx
```


> If Max is a known constant

Unfortunately it isn't. It's usually the width/height of a bitmap or a kernel or the size of a gradient lookup table.

**Comment by andersmelander** [comment]
@CuriousKit Btw, while I got your attention, I think your old home page has been infected with a trojan downloader. Might want to get that fixed :-)

**Comment by CuriousKit** [comment]
My old homepage? You mean my Garjen site? I kind of abandoned that over a decade ago, using it only as a temporary storage space.

**Comment by andersmelander** [comment]
Yes, that's the one. 
Kaspersky reports [Trojan-Downloader.JS.Iframe.dbj](https://threats.kaspersky.com/en/threat/Trojan-Downloader.JS.Iframe/) but for some reason a VirusTotal scan of the site doesn't catch it.
If I open the site in Firefox it attempts to start a download so there definitely is something bad going on.

**Comment by CuriousKit** [comment]
> Yes, that's the one. Kaspersky reports [Trojan-Downloader.JS.Iframe.dbj](https://threats.kaspersky.com/en/threat/Trojan-Downloader.JS.Iframe/) but for some reason a VirusTotal scan of the site doesn't catch it. If I open the site in Firefox it attempts to start a download so there definitely is something bad going on.

For now I've disabled the index page, so going to the site just lists all of the files.  Not the cleanest solution, but it's been such a long time since I've worked on the site, and it was a friend who last revamped it, so I'm not sure where to start looking (that and I'm extremely rusty on my PHP etc.)

**Comment by tomwiel** [comment]
> ...and also used by some gradient functions: https://graphics32.github.io/Docs/Units/GR32_ColorGradients/Classes/TCustomGradientPolygonFiller/Properties/WrapMode.htm

Typically, in spreadmodes (for gradient, texture) all pixels are repeated, just in reversed order. 
So it's not the triangle function. Tile size (and the number of tiles) does not change by the spreadmode.

**Comment by andersmelander** [comment]
> Typically, in spreadmodes (for gradient, texture) all pixels are repeated, just in reversed order.
> So it's not the triangle function. Tile size (and the number of tiles) does not change by the spreadmode.

I'm not sure what your point is. E.g. OpenGL's texture wrap mode concept is exactly the same as Graphics32's:
![billede](https://github.com/user-attachments/assets/210a567a-206a-4de7-a3e7-9987f71fe9ee)

If you plot the output of the `Mirror` function relative to the input value you will get a **triangle** wave, with a wavelength of 2*Max.
![billede](https://github.com/user-attachments/assets/d1c1d027-99ac-4bac-af32-92b8ebf9a8b3)

 The `Repeat` function produces a **sawtooth** with a wavelength of Max.
![billede](https://github.com/user-attachments/assets/5d11a023-22c5-4f5d-90ba-128da3d7713f)

And finally, the `Clamp` function:
![billede](https://github.com/user-attachments/assets/827146b5-86e9-4a07-b9e0-7a60243bc1ec)

**Comment by tomwiel** [comment]
With (max = 3):
012301230123  repeat (3 tiles, each 4 pixels), correct
012332100123  mirror old (3 tiles, each 4 pixels), correct (IMO)
012321012321  mirror new (more than 3 tiles), perfect triangle wave

**Comment by andersmelander** [comment]
> 012332100123 mirror old (3 tiles, each 4 pixels), correct (IMO)

Okay, now I get your meaning.

I guess my comparison with OpenGL wrap modes was badly chosen because for something like texture wrapping the old behavior was the correct one. However for the things that it was actually meant to be used for internally by Graphics32 the behavior was not correct or as-designed. For both gradients and resampling, duplicating the edge value/pixels would not produce the correct or expected result.

**Comment by tomwiel** [comment]
Here is an SVG example, with equal gradient vector size (tile size).
![pservers-grad-10-b](https://github.com/user-attachments/assets/693065ee-8cfe-443b-8d0c-9e6ccfd73dbe)

**Comment by andersmelander** [comment]
> Here is an SVG example, with equal gradient vector size (tile size).

Yes, SVG gradients are defined like that but that doesn't mean that all gradient implementations must be like that.
FWIW, Graphics32 predates SVG although Graphics32 didn't have gradients back then. The wrap mode was introduced for other purposes (e.g. resampling) and then later re-purposed for gradients.

If you can come up with a good reason why a gradient wrap mode, that behaves like SVG, is required then we can consider adding that as a new mode. It's not like it's a lot of work as we already have the code for it but it will cause some breakage as we will have to expand the `TWrapMode` enum and there is user code that relies on the number of elements in that type.

**Comment by lamdalili** [comment]
Currently the division is used to determine the direction of the ramp by checking the lowest bit, so it's possible to  acheive this by performing the modulo operation on the double of Max, then determining in which half the expected value lies.

```
 function Mirror(Value, Max:integer): integer;
 begin
  if Value < 0 then
    Value := -Value;
  Result := Value mod (Max * 2);
    if Result > Max then
       Result := Max * 2 - Result;
 end;
```

**Comment by andersmelander** [comment]
> Currently the division is used to determine the direction of the ramp by checking the lowest bit, so it's possible to acheive this by performing the modulo operation on the double of Max, then determining in which half the expected value lies.

Well, we get the division for free [when performing the modulus](https://www.felixcloutier.com/x86/idiv) or vice versa, so this doesn't really save anything and would still perform worse than the iterative version.

**Comment by andersmelander** [comment]
> [opencv](https://docs.opencv.org/4.x/d2/de8/group__core__array.html#ga209f2f4869e304c82d07739337eae7c5) supports two options: BORDER_REFLECT, BORDER_REFLECT_101

For reference here's one of the many (there are a lot) implementations of `BORDER_REFLECT` in OpenCV:
```C
// fedcba|abcdefgh|hgfedcb
#define EXTRAPOLATE(x, maxV) \
    { \
        (x) = min(((maxV)-1)*2-(x)+1, max((x),-(x)-1) ); \
    }
```
https://github.com/opencv/opencv/blob/53a5b85d9f6e2a58066f966d901c0e93ea1ebcf0/modules/imgproc/src/opencl/filterSep_singlePass.cl#L62-L66

and `BORDER_REFLECT_101`:
```C
// gfedcb|abcdefgh|gfedcba
#define EXTRAPOLATE(x, maxV) \
    { \
        (x) = min(((maxV)-1)*2-(x), max((x),-(x)) ); \
    }
```
https://github.com/opencv/opencv/blob/53a5b85d9f6e2a58066f966d901c0e93ea1ebcf0/modules/imgproc/src/opencl/filterSep_singlePass.cl#L68-L72

None of the definitions I found seemed to support more than +/-1 cycle but I didn't look that closely.

**Comment by lamdalili** [comment]
> Well, we get the division for free [when performing the modulus](https://www.felixcloutier.com/x86/idiv) or vice versa, so this doesn't really save anything and would still perform worse than the iterative version.

Yes I know but in PUREPASCAL mode there is no way to do that without performing the two operations this was was goal ,  the real question is how the function is used in the existing code I can guess that the optimization gain is extremely  marginal.

from  GR32_ColorGradients
```
FWrapProc(Round(FScale * Abs(FAngle + ArcTan2(Y, X))),

FWrapProc(Round(Sqrt(Sqr(X) + Sqr(Y)) * FScale), FLutMask)];

FWrapProc(Round((X - XOffset[0]) * Scale), Mask)
```

**Comment by andersmelander** [comment]
>  I can guess that the optimization gain is extremely marginal.

I'm not sure what optimization gain you mean but AFAIR the iterative solution was about twice as fast as the DivMod solution (asm versions, even more for the Pascal version).
For gradients this won't matter much as the bottleneck lies elsewhere, as you exemplified above, but for resampling and other pixel access it will.
Anyway, it's water under the bridge. As far as I'm concerned the problem has been resolved.

**Comment by andersmelander** [comment]
I have now reintroduced the old `wmMirror` wrap mode behavior as the new `wmReflect` wrap mode (in the **MirrorReflect** branch).
`wmReflect` behaves like OpenVC's `BORDER_REFLECT` and `wmMirror` behaves like `BORDER_REFLECT_101`.

I had to update a few of the examples that relied on the old declaration of `TWrapMode`. If the new wrap mode breaks too much in user code it can be disabled by undefining the `GR32_WRAPMODE_REFLECT` compiler define.

Interestingly, while I updated the existing gradient code to support the new wrap mode, I discovered that the gradient resamplers used the term **Reflect** to refer to what used to be the old `wmWrap` wrap mode (i.e. what is now the new `wmReflect` wrap mode).

With both `wmWrap` and `wmReflect` implemented, it also becomes easy to visualize why the old wrap implementation was incorrect for use with resamplers. Here's the Resamplers example showing the different edge handling (as it is now):
![billede](https://github.com/user-attachments/assets/010915ed-c0d7-434b-87e1-38b6af8e8ae6)
![billede](https://github.com/user-attachments/assets/f9c4b846-99cc-4f25-8d81-690872a600b2)
